File: test/test_telemetry_json.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stddef.h>

#include "telemetry.h"
#include "telemetry_serializer.h"
#include "room_state.h"

/* JSON wire-format regression harness.

   Emits one serialized schema-v3 telemetry payload per case on stdout, each
   prefixed with "CASE <n>". A companion python script (test_telemetry_json.py)
   pipes each blob through a real JSON parser (json.loads) and asserts the
   decoded values. This proves every Telemetry_Serialize() output is valid
   JSON — the substring-only tests could not catch malformed output. */

static void flush_case(int n, const TelemetrySnapshot *snap)
{
    uint8_t buf[TELEMETRY_SERIALIZED_MAX_SIZE];
    size_t written = 0;
    SerializeStatus s = Telemetry_Serialize(snap, buf, sizeof(buf), &written);
    if (s != SERIALIZE_OK)
    {
        printf("CASE %d\nSERIALIZE_ERROR\n", n);
        return;
    }
    buf[written] = '\0';
    printf("CASE %d\n%s", n, (const char *)buf);
}

int main(void)
{
    /* 1: all SCD41 values valid + illuminance. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 42;
        snap.uptime_ms = 10000;
        snap.captured_at_ms = 10000;
        snap.health = SYSTEM_HEALTH_OK;
        memcpy(snap.device_id,
               "\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10", 16);
        snap.room.illuminance_lux = 72.4f;
        snap.room.illuminance_valid = true;
        snap.room.co2_ppm = 1006.0f;
        snap.room.co2_valid = true;
        snap.room.scd41_temperature_c = 27.1f;
        snap.room.scd41_temperature_valid = true;
        snap.room.scd41_humidity_pct = 49.0f;
        snap.room.scd41_humidity_valid = true;
        flush_case(1, &snap);
    }

    /* 2: all SCD41 values invalid. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 43;
        snap.uptime_ms = 11000;
        snap.captured_at_ms = 11000;
        snap.health = SYSTEM_HEALTH_DEGRADED;
        snap.room.co2_valid = false;
        snap.room.scd41_temperature_valid = false;
        snap.room.scd41_humidity_valid = false;
        flush_case(2, &snap);
    }

    /* 3: mixed validity (co2 valid, T invalid, RH valid). */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 44;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = 500.0f;
        snap.room.co2_valid = true;
        snap.room.scd41_temperature_valid = false;
        snap.room.scd41_humidity_pct = 30.0f;
        snap.room.scd41_humidity_valid = true;
        flush_case(3, &snap);
    }

    /* 4: maximum legal CO2 = 40000. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 45;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = 40000.0f;
        snap.room.co2_valid = true;
        flush_case(4, &snap);
    }

    /* 5: CO2 over max (> 40000) -> must be serialized invalid. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 46;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = 40001.0f;
        snap.room.co2_valid = true;
        flush_case(5, &snap);
    }

    /* 6: negative CO2 (valid flag hacked true) -> invalid. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 47;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = -5.0f;
        snap.room.co2_valid = true;
        flush_case(6, &snap);
    }

    /* 7: CO2 = NaN -> invalid (never a numeric value). */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 48;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = NAN;
        snap.room.co2_valid = true;
        flush_case(7, &snap);
    }

    /* 8: CO2 = +Inf -> invalid. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 49;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = INFINITY;
        snap.room.co2_valid = true;
        flush_case(8, &snap);
    }

    /* 9: CO2 = -Inf -> invalid. */
    {
        TelemetrySnapshot snap;
        memset(&snap, 0, sizeof(snap));
        snap.sequence = 50;
        snap.health = SYSTEM_HEALTH_OK;
        snap.room.co2_ppm = -INFINITY;
        snap.room.co2_valid = true;
        flush_case(9, &snap);
    }

    return 0;
}

File: test/test_telemetry_json.py
#!/usr/bin/env python3
"""Real JSON-parser validation of schema-v3 telemetry serialization.

Reads the output of the test_telemetry_json host binary on stdin and validates
every serialized payload with a real JSON parser (json.loads). Asserts the
decoded values and the CO2 finite/range invalid semantics.

Usage: test_telemetry_json.py <path-to-test_telemetry_json-binary>  (or read via
run/CTest)  -- the binary path is argv[1]; default "test_telemetry_json".
"""
import json
import subprocess
import sys

BIN = sys.argv[1] if len(sys.argv) > 1 else "test_telemetry_json"

def case_block(out):
    """Yield (case_num, json_text) for each 'CASE n' block."""
    blocks = {}
    cur = None
    buf = []
    for line in out.splitlines(keepends=True):
        if line.startswith("CASE "):
            cur = line.strip().split()[1]
            buf = []
        else:
            buf.append(line)
        blocks[cur] = "".join(buf) if cur else ""
    return blocks

def main():
    proc = subprocess.run([BIN], capture_output=True, text=True)
    if proc.returncode != 0:
        print("FATAL: host binary failed rc=%d\n%s" % (proc.returncode, proc.stderr))
        return 1
    out = proc.stdout
    failures = 0
    cases = {}

    # split into CASE blocks by scanning lines
    cur = None
    cur_buf = []
    for line in out.splitlines(keepends=True):
        if line.startswith("CASE "):
            if cur is not None:
                cases[cur] = "".join(cur_buf)
            cur = line.strip().split()[1]
            cur_buf = []
        else:
            cur_buf.append(line)
    if cur is not None:
        cases[cur] = "".join(cur_buf)

    def check(cond, msg):
        nonlocal failures
        if cond:
            print("  PASS: %s" % msg)
        else:
            print("  FAIL: %s" % msg)
            failures += 1

    def parse_case(n):
        raw = cases.get(str(n))
        if raw is None:
            check(False, "case %d present" % n)
            return None
        if raw.strip() == "SERIALIZE_ERROR":
            check(False, "case %d serialized OK" % n)
            return None
        try:
            return json.loads(raw)
        except Exception as e:
            check(False, "case %d parses as JSON (%s)" % (n, e))
            return None

    print("=== JSON validity + parsed values ===")

    # Case 1: golden, all valid
    d = parse_case(1)
    if d is not None:
        check(d.get("schema") == 3, "case1 schema==3")
        bid = d.get("boot_id")
        check(isinstance(bid, str) and len(bid) == 16, "case1 boot_id length==16")
        check(isinstance(bid, str) and all(c in "0123456789abcdef" for c in bid),
              "case1 boot_id chars in [0-9a-f]")
        check(bid == "0123456789abcdef", "case1 boot_id == 0123456789abcdef")
        room = d.get("room", {})
        check(room.get("co2_ppm", {}).get("value") == 1006, "case1 co2.value==1006")
        check(room.get("co2_ppm", {}).get("state") == "valid", "case1 co2.state valid")
        check(abs(room.get("scd41_temperature_c", {}).get("value") - 27.1) < 0.01,
              "case1 scd41_temperature_c.value==27.1")
        check(abs(room.get("scd41_humidity_pct", {}).get("value") - 49.0) < 0.01,
              "case1 scd41_humidity_pct.value==49.0")
        check(abs(room.get("illuminance_lux", {}).get("value") - 72.4) < 0.01,
              "case1 illuminance_lux.value==72.4")
        check("value" in room.get("scd41_humidity_pct", {}), "case1 humidity has value")

    # Case 2: all invalid
    d = parse_case(2)
    if d is not None:
        room = d.get("room", {})
        check(room.get("co2_ppm", {}).get("state") == "invalid", "case2 co2 invalid")
        check("value" not in room.get("co2_ppm", {}), "case2 co2 has NO value (never 0)")
        check(room.get("scd41_temperature_c", {}).get("state") == "invalid", "case2 T invalid")
        check(room.get("scd41_humidity_pct", {}).get("state") == "invalid", "case2 RH invalid")
        check("value" not in room.get("scd41_temperature_c", {}), "case2 T no value")
        check("value" not in room.get("scd41_humidity_pct", {}), "case2 RH no value")

    # Case 3: mixed
    d = parse_case(3)
    if d is not None:
        room = d.get("room", {})
        check(room.get("co2_ppm", {}).get("value") == 500, "case3 co2==500")
        check(room.get("scd41_temperature_c", {}).get("state") == "invalid", "case3 T invalid")
        check(room.get("scd41_humidity_pct", {}).get("value") == 30, "case3 RH==30")

    # Case 4: max legal CO2
    d = parse_case(4)
    if d is not None:
        check(d.get("room", {}).get("co2_ppm", {}).get("value") == 40000, "case4 co2==40000 valid")
        check(d.get("room", {}).get("co2_ppm", {}).get("state") == "valid", "case4 co2 state valid")

    # Cases 5-9: out-of-range / non-finite CO2 must be invalid, never numeric 0
    for n, nm in [(5, "co2>40000"), (6, "co2 negative"), (7, "co2 NaN"),
                  (8, "co2 +Inf"), (9, "co2 -Inf")]:
        d = parse_case(n)
        if d is not None:
            cp = d.get("room", {}).get("co2_ppm", {})
            check(cp.get("state") == "invalid", "case%d %s -> invalid" % (n, nm))
            check("value" not in cp, "case%d %s -> no numeric value" % (n, nm))

    # Case 10: full-mask boot_id
    d = parse_case(10)
    if d is not None:
        bid = d.get("boot_id")
        check(isinstance(bid, str) and len(bid) == 16, "case10 boot_id length==16")
        check(bid == "ffffffffffffffff", "case10 boot_id == ffffffffffffffff")
        check(all(c in "0123456789abcdef" for c in bid), "case10 boot_id chars in [0-9a-f]")

    # Golden parsed assertions
    if parse_case(1) is not None:
        print("  golden parsed: schema=%d co2=%s state=%s" % (
            parse_case(1).get("schema"),
            parse_case(1).get("room", {}).get("co2_ppm", {}).get("value"),
            parse_case(1).get("room", {}).get("co2_ppm", {}).get("state")))

    sys.exit(1 if failures else 0)

if __name__ == "__main__":
    main()